zebra-chain/src/orchard/shielded_data.rs

```diff
@@ -0,0 +1,128 @@
+//! Orchard shielded data for `V5` `Transaction`s.
+
+use crate::{
+    amount::Amount,
+    block::MAX_BLOCK_BYTES,
+    orchard::{tree, Action},
+    primitives::{
+        redpallas::{Binding, Signature, SpendAuth},
+        Halo2Proof,
+    },
+    serialization::{
+        AtLeastOne, SerializationError, TrustedPreallocate, ZcashDeserialize, ZcashSerialize,
+    },
+};
+
+use byteorder::{ReadBytesExt, WriteBytesExt};
+
+use std::{
+    cmp::{Eq, PartialEq},
+    fmt::Debug,
+    io,
+};
+
+/// A bundle of [`Action`] descriptions and signature data.
+#[derive(Clone, Debug, PartialEq, Eq, Deserialize, Serialize)]
+pub struct ShieldedData {
+    /// The orchard flags for this transaction.
+    pub flags: Flags,
+    /// The net value of Orchard spends minus outputs.
+    pub value_balance: Amount,
+    /// The shared anchor for all `Spend`s in this transaction.
+    pub shared_anchor: tree::Root,
+    /// The aggregated zk-SNARK proof for all the actions in this transaction.
+    pub proof: Halo2Proof,
+    /// The Orchard Actions.
+    pub actions: AtLeastOne<AuthorizedAction>,
+    /// A signature on the transaction `sighash`.
+    pub binding_sig: Signature<Binding>,
+}
+
+/// An authorized action description.
+///
+/// Every authorized Orchard `Action` must have a corresponding `SpendAuth` signature.
+#[derive(Clone, Debug, PartialEq, Eq, Deserialize, Serialize)]
+pub struct AuthorizedAction {
+    /// The action description of this Action.
+    pub action: Action,
+    /// The spend signature.
+    pub spend_auth_sig: Signature<SpendAuth>,
+}
+
+impl AuthorizedAction {
+    /// Split out the action and the signature for V5 transaction
+    /// serialization.
+    pub fn into_parts(self) -> (Action, Signature<SpendAuth>) {
+        (self.action, self.spend_auth_sig)
+    }
+
+    // Combine the action and the spend auth sig from V5 transaction
+    /// deserialization.
+    pub fn from_parts(action: Action, spend_auth_sig: Signature<SpendAuth>) -> AuthorizedAction {
+        AuthorizedAction {
+            action,
+            spend_auth_sig,
+        }
+    }
+}
+
+/// The size of a single Action
+///
+/// Actions are 5 * 32 + 580 + 80 bytes so the total size of each Action is 820 bytes.
+/// [7.5 Action Description Encoding and Consensus][ps]
+///
+/// [ps] https://zips.z.cash/protocol/nu5.pdf#actionencodingandconsensus
+pub const ACTION_SIZE: u64 = 5 * 32 + 580 + 80;
+
+impl TrustedPreallocate for Action {
+    fn max_allocation() -> u64 {
+        (MAX_BLOCK_BYTES - 1) / ACTION_SIZE
+    }
+}
```

Comment:
Please use `884` bytes as the size limit to calculate `Action::max_allocation`.

According to the spec, `Action`s are `5 * 32 + 580 + 80` bytes and `Signature<SpendAuth>` is `64` bytes, so the total size of each `Action` + `Signature<SpendAuth>` is `884` bytes. (Since there must be a `Signature<SpendAuth>` for each `Action`, we use both their sizes to calculate the limit.)

So we can limit the number of `Action`s in a block or transaction to `(MAX_BLOCK_BYTES - 1) / 884`. (The length is at least 1 byte.)

The examples in https://github.com/ZcashFoundation/zebra/pull/2116#discussion_r632147435 show how to handle the size of extra required fields.

Reply:
By https://github.com/ZcashFoundation/zebra/pull/2116#discussion_r632570639 we need to have `TrustedPreallocate` for `Signature<SpendAuth>` defined independently from `Action` so we can deserialize.

The suggestions are to increase the size of `Action` (as it always come together with a `Signature<SpendAuth>`) even if we still need to have `Signature<SpendAuth>` defined apart, then remove the tests for `Signature<SpendAuth>`.

Please clarify. Thanks.

Reply:
You're right, I'm working off a set of examples that don't match what we're doing here.

Use `SpendPrefixInTransactionV5` and `redjubjub::Signature<SpendAuth>` as examples instead. You should be able to copy-paste the code, and then change the types, sizes, and constant names.

https://github.com/ZcashFoundation/zebra/blob/2ea1d8624d1d72abdf582953f4f1133efe9eea4f/zebra-chain/src/sapling/spend.rs#L254

https://github.com/ZcashFoundation/zebra/blob/2ea1d8624d1d72abdf582953f4f1133efe9eea4f/zebra-chain/src/sapling/tests/preallocate.rs#L35

https://github.com/ZcashFoundation/zebra/blob/2ea1d8624d1d72abdf582953f4f1133efe9eea4f/zebra-chain/src/sapling/tests/preallocate.rs#L75

You might want to make the `max_allocation_is_big_enough` function `pub(crate)` and re-use it.